plugin/pulpcore/plugin/changeset/model.py

```diff
@@ -198,10 +295,58 @@ def save(self):
         Update the DB model to store the downloaded file.
         Then, save in the DB.
         """
-        if self.path:
-            self.model.file = File(open(self.path, mode='rb'))
-            self.model.downloaded = True
-        super(RemoteArtifact, self).save()
+        artifact = None
+
+        if not self.fetched:
+            if self._path:  # downloaded
+                try:
+                    with transaction.atomic():
```

Comment:
All of this is being done in 1 main transaction started by the `PendingContent.save()` which invokes each `PendingArtifact.save()`.  The nested transactions here are really acting as _checkpoints_.  Without them, the `IntegrityError` would put the main transaction in an error state and the `get()` in the _except_ block would be rejected by django.

The code here is just dealing with duplicates created by race conditions.  To properly deal with the situation you described with a referenced row being deleted by another process (also a race condition), this code would need a way to differentiate between the two errors.  Unfortunately, except for parsing the description of the IntegrityError (ugh), I don't see a way to do that.  Django should raise a specific exception for dups but best I can tell (and googled), it does not.  If it did, I think we could deal with that by catching an exception in `PendingContent.save()` and retrying the entire transaction.